--- a/agent/snmp_agent.c
+++ b/agent/snmp_agent.c
@@ -2995,6 +2995,10 @@
     return asp->status;
 }
 
+
+void (* pre_call_handle_var_requests)( netsnmp_agent_session *asp) = NULL;
+void (* post_call_handle_var_requests)( netsnmp_agent_session *asp) = NULL;
+
 int
 handle_var_requests(netsnmp_agent_session *asp)
 {
@@ -3005,6 +3009,9 @@
     asp->reqinfo->asp = asp;
     asp->reqinfo->mode = asp->mode;
 
+	if(pre_call_handle_var_requests != NULL)
+		pre_call_handle_var_requests(asp);
+
     /*
      * now, have the subtrees in the cache go search for their results 
      */
@@ -3080,6 +3087,10 @@
         }
     }
 
+
+	if(post_call_handle_var_requests != NULL)
+		post_call_handle_var_requests(asp);
+
     return final_status;
 }
 
